go-controller/pkg/ovn/address_set_sync.go

```diff
@@ -0,0 +1,424 @@
+		reflist, ok := syncer.replaceRefLRPs[lrp.UUID]
+		if !ok {
+			reflist = refsList[*nbdb.LogicalRouterPolicy]{
+				dbobj:   lrp,
+				replace: map[string]string{},
+			}
+			syncer.replaceRefLRPs[lrp.UUID] = reflist
+		}
+		reflist.replace[as.Name] = newHash
+	}
+
+	if staleAS {
+		deleteOps, err := syncer.getDeleteASOps(acls, qoses, lrps, as)
+		if err != nil {
+			return fmt.Errorf("failed to get delete address set ops: %v", err)
+		}
+		syncer.staleASOps[oldName] = staleASInfo{deleteOps, ops}
+	} else {
+		syncer.newASs[oldName] = true
+		syncer.ops = append(syncer.ops, ops...)
+	}
+	return nil
+}
+
+func (syncer *addressSetsSyncer) SyncAddressSets() error {
+	// stale address sets don't have controller ID
+	p := libovsdbops.GetNoOwnerPredicate[*nbdb.AddressSet]()
+	addrSetList, err := libovsdbops.FindAddressSetsWithPredicate(syncer.nbClient, p)
+	if err != nil {
+		return fmt.Errorf("failed to find stale address sets: %v", err)
+	}
+
+	for _, as := range addrSetList {
+		if err = syncer.updateOpsForAS(as); err != nil {
+			return err
+		}
+	}
+	// update refs
+	aclsToUpdate := []*nbdb.ACL{}
+	for _, replaceRef := range syncer.replaceRefACLs {
+		acl := replaceRef.dbobj
+		for oldHash, newHash := range replaceRef.replace {
+			acl.Match = strings.ReplaceAll(acl.Match, "$"+oldHash, "$"+newHash)
+		}
+		aclsToUpdate = append(aclsToUpdate, acl)
+	}
+	ops, err := libovsdbops.CreateOrUpdateACLsOps(syncer.nbClient, nil, aclsToUpdate...)
+	if err != nil {
+		return fmt.Errorf("failed to get update acl ops: %v", err)
+	}
+
+	qosesToUpdate := []*nbdb.QoS{}
+	for _, replaceRef := range syncer.replaceRefQoSes {
+		qos := replaceRef.dbobj
+		for oldHash, newHash := range replaceRef.replace {
+			qos.Match = strings.ReplaceAll(qos.Match, "$"+oldHash, "$"+newHash)
+		}
+		qosesToUpdate = append(qosesToUpdate, qos)
+	}
+	ops, err = libovsdbops.CreateOrUpdateQoSesOps(syncer.nbClient, ops, qosesToUpdate...)
+	if err != nil {
+		return fmt.Errorf("failed to get update qos ops: %v", err)
+	}
+
+	lrpsToUpdate := []*nbdb.LogicalRouterPolicy{}
+	for _, replaceRef := range syncer.replaceRefLRPs {
+		lrp := replaceRef.dbobj
+		for oldHash, newHash := range replaceRef.replace {
+			lrp.Match = strings.ReplaceAll(lrp.Match, "$"+oldHash, "$"+newHash)
+		}
+		lrpsToUpdate = append(lrpsToUpdate, lrp)
+	}
+	ops, err = libovsdbops.CreateOrUpdateLogicalRouterPoliciesOps(syncer.nbClient, ops, lrpsToUpdate...)
+	if err != nil {
+		return fmt.Errorf("failed to get update LRPs ops: %v", err)
+	}
+	syncer.ops = append(syncer.ops, ops...)
+
+	// delete stale address sets
+	for asName, staleAS := range syncer.staleASOps {
+		if syncer.newASs[asName] {
+			// newer version for the same address set is found, delete old
+			syncer.ops = append(syncer.ops, staleAS.deleteOps...)
+		} else {
+			// no newer version, update stale AS
+			syncer.ops = append(syncer.ops, staleAS.updateOps...)
+		}
+	}
+	_, err = libovsdbops.TransactAndCheck(syncer.nbClient, syncer.ops)
```

Comment:
I think that updating every AS of a system in a single transaction might be our biggest transaction yet and I am not really sure if it will overflow something down the line. I am not sure either that we can do early scale testing of this, so we might receive feedback late in the cycle.
Because of this, it might be worth considering:
* Prepare the code a bit better to encapsulate the migration of a single address set. In case we need to do it lazily from the address set factory in the future.
* Prepare the code a bit better to batch the migration by a certain number of AS at a time that we can easily tweak once we receive feedback from scale testing.

Or maybe there are reasons to believe this probably will not be a problem at scale that I am missing?

Reply:
definitely need to consider this: I think it should be easier to do that per-address set, and it will simplify the code, but also that may take a very long time in case of a big cluster. So performance-wise batching should be the best option I guess, but it may complicate the code a bit more

Reply:
I implemented batching, set the number of address sets per transaction to 50 (random number, may be around 150 ops, hopefully libovsdb can handle this)

Reply:
Cool, thanks!

Could we possibly allow `txnBatchSize` to be 0 for no batching?

And maybe the default? I just wanted to be prepared. Not a clear idea though, so I am ok with 50 too.

I would probably be cleaner to have `addressSetBatchSync` type or method, to have a cleaner `SyncAddressSets` method. But this is a nit.

Reply:
I am not sure if we need no batching in general, having some limit for a transaction without any predictable limit sound reasonable for me, wdyt? 
With the latest changes the inner part of the batching part is so small, I don't think it makes sense to make it another function (which definitely made sense for the previous version)